commcare_connect/opportunity/migrations/0081_create_duplicated_completed_works.py

```diff
@@ -0,0 +1,89 @@
+# Generated by Django 4.2.5 on 2025-07-14 06:32
+
+from collections import defaultdict
+import logging
+
+from django.db import migrations, models
+from django.db.models import F, Window
+from django.db.models.functions import Rank
+
+
+def create_deduplicated_completed_works(apps, schema_editor):
+    UserVisit = apps.get_model("opportunity.UserVisit")
+    CompletedWork = apps.get_model("opportunity.CompletedWork")
+    user_visits = (
+        UserVisit.objects.annotate(
+            rank=Window(
+                expression=Rank(),
+                order_by=F("visit_date").desc(),
+                partition_by=[
+                    F("entity_id"),
+                    F("opportunity_access_id"),
+                    F("deliver_unit_id"),
+                    F("completed_work_id"),
+                ],
+            )
+        )
+        .order_by("entity_id")
+        .filter(rank__gt=1, completed_work__isnull=False)
+    )
+
+    logging.info(f"{user_visits.count()} duplicate user visits found.")
+
+    completed_work_user_visits_map = defaultdict(list)
+    completed_work_details_map = defaultdict(dict)
+
+    for user_visit in user_visits:
+        entity_id = user_visit.entity_id
+        access_id = user_visit.opportunity_access_id
+        payment_unit_id = user_visit.deliver_unit.payment_unit_id
+        rank = user_visit.rank
+        key = entity_id, access_id, payment_unit_id, rank
+        completed_work_user_visits_map[key].append(user_visit)
+        completed_work_details_map[key].update(
+            {
+                "entity_id": entity_id,
+                "entity_name": user_visit.entity_name,
+                "opportunity_access_id": access_id,
+                "payment_unit_id": payment_unit_id,
+            }
+        )
+
+    to_update = []
+    for key, details in completed_work_details_map.items():
+        completed_work = CompletedWork(**details)
```

Comment:
what status do these have? is there a way to determine a "correct" status based on the underlying visits?

Reply:
These did not have a status. I have added a Duplicate status for completed works now which will be assigned here.